### PR TITLE
Implement SslPolicies for TargetHttpsProxy

### PR DESCRIPTION
**Pull request by spencerhance**
This is the main feature PR for SslPolicies

This relies on a few other PRs, but can be reviewed in parallel

### CODE REVIEW

File: pkg/loadbalancers/target_proxies.go [hunk 1]
```diff
@@ -190,3 +208,33 @@ func (l *L7) getSslCertLinkInUse() ([]string, error) {
 
 	return proxy.SslCertificates, nil
 }
+
+func (l *L7) GetSslPolicyLink() (*string, error) {
+	var link string
```
Comment (rramkumar1) [inline pkg/loadbalancers/target_proxies.go]: I think this method can be detached and unexported. Just pass in the runtime info

File: pkg/loadbalancers/target_proxies.go [hunk 2]
```diff
@@ -190,3 +208,33 @@ func (l *L7) getSslCertLinkInUse() ([]string, error) {
...
+
+	if l.runtimeInfo.FrontendConfig == nil {
+		return nil, nil
+	}
+
+	policyName := l.runtimeInfo.FrontendConfig.Spec.SslPolicy
+	if policyName == nil {
+		return nil, nil
+	}
+
```
Comment (rramkumar1) [inline pkg/loadbalancers/target_proxies.go]: remove this space

File: pkg/loadbalancers/target_proxies.go [hunk 3]
```diff
@@ -190,3 +208,33 @@ func (l *L7) getSslCertLinkInUse() ([]string, error) {
...
+	if *policyName == "" {
+		return &link, nil
+	}
+
+	key, err := l.CreateKey(*policyName)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (rramkumar1) [inline pkg/loadbalancers/target_proxies.go]: remove this space

File: pkg/loadbalancers/target_proxies.go (hunk 1 above)
Comment (spencerhance) [reply pkg/loadbalancers/target_proxies.go]: I can definitely unexport it, but I would also need to pass in the L7 anyway to access CreateKey()

Comment (rramkumar1) [reply pkg/loadbalancers/target_proxies.go]: Ack. Also this needs a unit test.

File: pkg/loadbalancers/target_proxies.go [hunk 4]
```diff
@@ -171,8 +172,25 @@ func (l *L7) checkHttpsProxy() (err error) {
 		if err := composite.SetSslCertificateForTargetHttpsProxy(l.cloud, key, proxy, sslCertURLs); err != nil {
 			return err
 		}
+	}
+
+	if flags.F.EnableFrontendConfig {
```
Comment (rramkumar1) [inline pkg/loadbalancers/target_proxies.go]: is there an existing unit test you can integrate this new logic into ?

File: pkg/loadbalancers/target_proxies.go (hunk 1 above)
Comment (rramkumar1) [reply pkg/loadbalancers/target_proxies.go]: Ack. Also needs a unit test.

**Comment by spencerhance** [comment]
@rramkumar1 Added a bunch of unit tests, can you take another look?

File: pkg/loadbalancers/loadbalancers_test.go [hunk 5]
```diff
@@ -97,6 +99,18 @@ func newTestJig(t *testing.T) *testJig {
 		tp.SslCertificates = request.SslCertificates
 		return nil
 	}
+	mockGCE.MockTargetHttpsProxies.SetSslPolicyHook = func(ctx context.Context, key *meta.Key, ref *compute.SslPolicyReference, proxies *cloud.MockTargetHttpsProxies) error {
```
Comment (rramkumar1) [inline pkg/loadbalancers/loadbalancers_test.go]: Is this hook special? Can it be added to k8s_cloud_provider mock.go instead? 

If its too much work to re-vendor then I suppose we can have this here but would be good to upstream

File: pkg/utils/utils.go [hunk 6]
```diff
@@ -439,3 +439,8 @@ func IsLegacyL4ILBService(svc *api_v1.Service) bool {
 	}
 	return false
 }
+
+// NewString returns a pointer to the provider string literal
```
Comment (rramkumar1) [inline pkg/utils/utils.go]: provided

File: pkg/utils/utils.go [hunk 7]
```diff
@@ -439,3 +439,8 @@ func IsLegacyL4ILBService(svc *api_v1.Service) bool {
...
+func NewString(s string) *string {
```
Comment (rramkumar1) [inline pkg/utils/utils.go]: NewStringPointer

File: pkg/loadbalancers/loadbalancers_test.go [hunk 8]
```diff
@@ -973,6 +987,154 @@ func TestCreateBothLoadBalancers(t *testing.T) {
...
+
+func TestGetSslPolicyLink(t *testing.T) {
+	t.Parallel()
+	j := newTestJig(t)
+
+	testCases := []struct {
+		desc string
+		fc   *frontendconfigv1beta1.FrontendConfig
+		want *string
+	}{
```
Comment (rramkumar1) [inline pkg/loadbalancers/loadbalancers_test.go]: Do you need a case for frontend config with empty string ssl policy?

File: pkg/loadbalancers/loadbalancers_test.go [hunk 9]
```diff
@@ -973,6 +987,154 @@ func TestCreateBothLoadBalancers(t *testing.T) {
...
+			want:  "",
+		},
+		{
+			desc:  "frontendconfig with ssl policy",
+			fc:    &frontendconfigv1beta1.FrontendConfig{Spec: frontendconfigv1beta1.FrontendConfigSpec{SslPolicy: utils.NewString("test-policy")}},
+			proxy: &composite.TargetHttpsProxy{Name: "test-proxy-2"},
+			want:  "global/sslPolicies/test-policy",
+		},
+		{
+			desc:  "proxy with different ssl policy",
```
Comment (rramkumar1) [inline pkg/loadbalancers/loadbalancers_test.go]: Is this a duplicate of the test case below?

File: pkg/loadbalancers/loadbalancers_test.go [hunk 10]
```diff
@@ -973,6 +987,154 @@ func TestCreateBothLoadBalancers(t *testing.T) {
...
+			want:  "global/sslPolicies/test-policy",
+		},
+		{
+			desc:  "remove ssl policy",
+			fc:    &frontendconfigv1beta1.FrontendConfig{Spec: frontendconfigv1beta1.FrontendConfigSpec{SslPolicy: utils.NewString("")}},
+			proxy: &composite.TargetHttpsProxy{Name: "test-proxy-5", SslPolicy: "global/sslPolicies/wrong-policy"},
+			want:  "",
+		},
+	}
+
```
Comment (rramkumar1) [inline pkg/loadbalancers/loadbalancers_test.go]: You need a case where a proxy has a policy and then the frontend config policy string is nil.

File: pkg/loadbalancers/loadbalancers_test.go (hunk 5 above)
Comment (spencerhance) [reply pkg/loadbalancers/loadbalancers_test.go]: Can I do this in a follow up and add a TODO here?  Re-vendoring always takes like a couple days to get all the approvals.

File: pkg/loadbalancers/loadbalancers_test.go (hunk 10 above)
Comment (spencerhance) [reply pkg/loadbalancers/loadbalancers_test.go]: Done

File: pkg/loadbalancers/loadbalancers_test.go (hunk 9 above)
Comment (spencerhance) [reply pkg/loadbalancers/loadbalancers_test.go]: Done

File: pkg/loadbalancers/loadbalancers_test.go (hunk 8 above)
Comment (spencerhance) [reply pkg/loadbalancers/loadbalancers_test.go]: Done

File: pkg/utils/utils.go (hunk 7 above)
Comment (spencerhance) [reply pkg/utils/utils.go]: Done

File: pkg/utils/utils.go (hunk 6 above)
Comment (spencerhance) [reply pkg/utils/utils.go]: Done

**Comment by rramkumar1** [comment]
Be sure to send a PR for the k8s_cloud_provider change

/lgtm

**Pull request merged**